### PR TITLE
feat: Add attachable CloudFront OAI and OAC policies

### PR DESCRIPTION
**Pull request by gw0**
## Description

Adds attachable CloudFront Origin Access Identities (deprecated by AWS) and CloudFront Origin Access Identities policies.

## Motivation and Context

An AWS S3 bucket is often connected with an AWS CloudFront distribution and configuring a securely authenticated connection should be trivial. Workarounds such as https://github.com/terraform-aws-modules/terraform-aws-cloudfront/blob/master/examples/complete/main.tf#L294-L327 should not be needed.

## Breaking Changes

None.

## How Has This Been Tested?
- [ ] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [ ] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [ ] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by gw0** [comment]
Is there ay reason why this should not be part of the TF module?

**Comment by gw0** [comment]
.

**Comment by gw0** [comment]
@antonbabenko This TF module already supports off-the-shelf policies for other AWS services. Why not also support AWS CloudFront?

**Comment by tculp** [comment]
unstale

File: main.tf [hunk 1]
```diff
@@ -763,6 +765,66 @@ data "aws_iam_policy_document" "access_log_delivery" {
...
+    resources = ["${aws_s3_bucket.this[0].arn}/*"]
+    principals {
+      type        = "AWS"
+      identifiers = var.cloudfront_oai_iam_arns
+    }
+  }
+
+  statement {
+    sid = "AWSCloudFrontOAIList"
+    actions   = ["s3:ListBucket"]
```
Comment (antonbabenko) [inline main.tf:784]: `s3:ListBucket` doesn't seem to be needed according to the docs - https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html#private-content-restricting-access-to-s3-oai

File: main.tf [hunk 2]
```diff
@@ -763,6 +765,66 @@ data "aws_iam_policy_document" "access_log_delivery" {
   }
 }
 
+# Grant read and list access to CloudFront Origin Access Identities (deprecated by AWS)
+data "aws_iam_policy_document" "cloudfront_oai_read_policy" {
+  count = local.create_bucket && var.attach_cloudfront_oai_read_policy ? 1 : 0
+
+  statement {
+    sid = "AWSCloudFrontOAIRead"
```
Comment (antonbabenko) [inline main.tf:773]: Please use the same `sid` as the one you are copying it from the documentation (e.g. `PolicyForCloudFrontPrivateContent`).

Comment (antonbabenko) [reply main.tf:773]: Let's change this statement to support `write` actions as well. Let's use boolean control named `var.attach_cloudfront_oai_policy` and allow parameterization of IAM actions used by OAI via a separate variable:
```
cloudfront_oai_iam_actions = ["s3:GetObject", "s3:PutObject"]
```

The default value is of `cloudfront_oai_iam_actions` is `["s3:GetObject"]`

File: main.tf [hunk 3]
```diff
@@ -763,6 +765,66 @@ data "aws_iam_policy_document" "access_log_delivery" {
   }
 }
 
+# Grant read and list access to CloudFront Origin Access Identities (deprecated by AWS)
+data "aws_iam_policy_document" "cloudfront_oai_read_policy" {
```
Comment (antonbabenko) [inline main.tf:769]:
```suggestion
data "aws_iam_policy_document" "cloudfront_oai_policy" {
```

File: main.tf [hunk 4]
```diff
@@ -763,6 +765,66 @@ data "aws_iam_policy_document" "access_log_delivery" {
...
+    resources = [aws_s3_bucket.this[0].arn]
+    principals {
+      type        = "AWS"
+      identifiers = var.cloudfront_oai_iam_arns
+    }
+  }
+}
+
+# Grant read and list access to CloudFront Origin Access Controls
+data "aws_iam_policy_document" "cloudfront_oac_read_policy" {
```
Comment (antonbabenko) [inline main.tf:794]:
```suggestion
data "aws_iam_policy_document" "cloudfront_oac_policy" {
```

File: main.tf [hunk 5]
```diff
@@ -763,6 +765,66 @@ data "aws_iam_policy_document" "access_log_delivery" {
...
+  count = local.create_bucket && var.attach_cloudfront_oac_read_policy ? 1 : 0
+
+  statement {
+    sid       = "AWSCloudFrontOACRead"
+    actions   = ["s3:GetObject"]
```
Comment (antonbabenko) [inline main.tf:799]: Let's apply the same flexibility in the support for read/write actions as I have described above for OAI.

**Review by antonbabenko — CHANGES_REQUESTED** [review]
Looks pretty good. There are a few comments about how to extend it to support read and write ops.

Also, please take a look at the examples in this module AND in the [cloudfront module](https://github.com/terraform-aws-modules/terraform-aws-cloudfront/blob/master/examples/complete/main.tf#L294-L327).

**Comment by tculp** [comment]
unstale

**Pull request closed without merging**